odoo/addons/base/models/ir_binary.py

```diff
@@ -0,0 +1,162 @@
+import logging
+import os.path
+import werkzeug.http
+from datetime import datetime
+from mimetypes import guess_extension
+from werkzeug.exceptions import NotFound
+
+from odoo import models
+from odoo.exceptions import UserError, MissingError
+from odoo.http import Stream, request
+from odoo.tools import file_open
+from odoo.tools.image import image_process, image_guess_size_from_field_name
+from odoo.tools.mimetypes import guess_mimetype
+
+
+DEFAULT_PLACEHOLDER_PATH = 'web/static/img/placeholder.png'
+_logger = logging.getLogger(__name__)
+
+
+def user_error_as_not_found(func):
+    def user_error_as_not_found_(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except UserError as exc:
+            raise NotFound() from exc
+    return user_error_as_not_found_
+
+
+class IrBinary(models.AbstractModel):
+    _name = 'ir.binary'
+    _description = "File streaming helper model for controllers"
+
+    def _find_record(
+            self, xmlid=None, res_model='ir.attachment', res_id=None,
+            access_token=None,
+    ):
+        record = None
+        if xmlid:
+            record = self.env.ref(xmlid, False)
+        elif res_id is not None and res_model in self.env:
+            record = self.env[res_model].browse(res_id).exists()
+        if not record:
+            raise MissingError(f"No record found for xmlid={xmlid}, res_model={res_model}, id={res_id}")
+
+        if record._name == 'ir.attachment':
+            record = record.validate_access(access_token)
+
+        # We have prefetched some fields of record. In order to check
+        # access on record, we have to invalidate its cache first.
+        if not record.env.su:
+            record._cache.clear()
+
+        record.check_access_rights('read')
```

Comment:
may be move the check of access in `else` after `if record._name == 'ir.attachment':` because the attachment is already checked.

Reply:
I need to clear the cache first and to always do it (although that code is kinda cargo-cult) so ye check-access cannot be in an else block (I think...)